// stillat-blade-parser/out/formatting/prettier/plugin.js
"use strict";
const bladeDocument_1 = require("../../document/bladeDocument");
const phpParserPhpValidator_1 = require("../../parser/php/phpParserPhpValidator");
const optionDiscovery_1 = require("../optionDiscovery");
const prettierDocumentFormatter_1 = require("./prettierDocumentFormatter");
const utils_1 = require("./utils");
let prettierOptions, transformOptions, parserOptions;
let bladeOptions = null;
const plugin = {
    languages: [
        {
            name: "Blade",
            parsers: ["blade"],
            extensions: [".blade.php"],
            vscodeLanguageIds: ["blade"],
        }
    ],
    parsers: {
        blade: {
            parse: function (text, _, options) {
                (0, utils_1.setOptions)(options);
                prettierOptions = options;
                bladeOptions = (0, optionDiscovery_1.getEnvSettings)(__dirname);
                transformOptions = bladeOptions;
                parserOptions = bladeOptions;
                transformOptions.tabSize = (0, utils_1.getHtmlOptions)().tabWidth;
                const document = new bladeDocument_1.BladeDocument();
                document.getParser()
                    .withParserOptions(parserOptions)
                    .withPhpValidator(new phpParserPhpValidator_1.PhpParserPhpValidator());
                return document.loadString(text);
            },
            locStart: () => 0,
            locEnd: () => 0,
            astFormat: "blade",
        },
    },
    printers: {
        blade: {
            print(path) {
                const doc = path.stack[0];
                return (new prettierDocumentFormatter_1.PrettierDocumentFormatter(prettierOptions, transformOptions)).formatDocument(doc);
            }
        }
    },
    defaultOptions: {
        tabWidth: 4,
    },
};
module.exports = plugin;
